mobile/src/main/java/org/openhab/habdroid/ui/WidgetAdapter.kt

```diff
@@ -825,42 +855,89 @@ class WidgetAdapter(
             labelView.isVisible = showLabelAndIcon
             iconView.isVisible = showLabelAndIcon
 
-            val mappings = widget.mappings.filter { it.column != 0 && it.row != 0 }
-            spareViews.addAll(table.children.map { it as? MaterialButton }.filterNotNull())
-            table.removeAllViews()
```

Comment:
@jimtng I'm currently looking at the code and am preparing some simplifications, and am wondering: was there a specific reason for this change (kill off `spareViews`, remove table children selectively)? I am asking because
- `spareViews` is an optimization for only creating new views if absolutely needed (it's a relatively expensive operation, as it involves e.g. XML parsing)
- removing table children selectively avoids a relatively cheap operation at the expense of much larger tracking overhead (the `buttonViews` map and its management isn't needed without it)

If there's no hard reason my proposal would be restoring the existing code. FWIW, my (currently untested) simplification proposal can be found [here](https://github.com/maniac103/openhab.android/commit/dfb20a06c342b1fa0ab45df54440688074ceec56)

Reply:
The whole aim of this exercise is to want to preserve the same buttonView object for the same grid location, and not calling `table.removeAllViews()`. When removeAllViews is called, the buttonView got wiped out, and that's why onTouch/Release tracking is gone. Touch down -> sends a command -> receives widget update -> calls bind() -> removeAllViews -> oops that buttonView waiting for the touch release just disappeared.

Thanks for your proposed code. I'll study it and get back to you on this.

Reply:
OK, got it, thanks. Updated proposal [here](https://github.com/maniac103/openhab.android/commit/efc14a3a263cb038d15262a3d01ee3047adfb986).

Reply:
I've had a look through your proposed code. I love that callback trick!

However, I think the removeAllViews will mess up with the touch release tracking. I tried your code and got the following exception

```
ACRA caught a IllegalStateException for org.openhab.habdroid.beta
java.lang.IllegalStateException: The specified child already has a parent. You must call removeView() on the child's parent first.
	at android.view.ViewGroup.addViewInner(ViewGroup.java:5275)
	at android.view.ViewGroup.addView(ViewGroup.java:5104)
	at android.view.ViewGroup.addView(ViewGroup.java:5076)
	at org.openhab.habdroid.ui.WidgetAdapter$ButtongridViewHolder.bind(WidgetAdapter.kt:898)
	at org.openhab.habdroid.ui.WidgetAdapter.onBindViewHolder(WidgetAdapter.kt:260)
	at org.openhab.habdroid.ui.WidgetAdapter.onBindViewHolder(WidgetAdapter.kt:128)
	at androidx.recyclerview.widget.RecyclerView$Adapter.onBindViewHolder(RecyclerView.java:7747)
	at androidx.recyclerview.widget.RecyclerView$Adapter.bindViewHolder(RecyclerView.java:7847)
	at androidx.recyclerview.widget.RecyclerView$Recycler.tryBindViewHolderByDeadline(RecyclerView.java:6646)
	at androidx.recyclerview.widget.RecyclerView$Recycler.tryGetViewHolderForPositionByDeadline(RecyclerView.java:6917)
	at androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:6757)
	at androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:6753)
	at androidx.recyclerview.widget.LinearLayoutManager$LayoutState.next(LinearLayoutManager.java:2362)
	at androidx.recyclerview.widget.LinearLayoutManager.layoutChunk(LinearLayoutManager.java:1662)
	at androidx.recyclerview.widget.LinearLayoutManager.fill(LinearLayoutManager.java:1622)
	at androidx.recyclerview.widget.LinearLayoutManager.onLayoutChildren(LinearLayoutManager.java:687)
	at androidx.recyclerview.widget.RecyclerView.dispatchLayoutStep2(RecyclerView.java:4645)
	at androidx.recyclerview.widget.RecyclerView.dispatchLayout(RecyclerView.java:4348)
	at androidx.recyclerview.widget.RecyclerView.onLayout(RecyclerView.java:4919)
	at android.view.View.layout(View.java:24421)
	at android.view.ViewGroup.layout(ViewGroup.java:6440)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
                                                                                                    	at android.view.View.layout(View.java:24421)
```

Reply:
I commented before seeing your updated proposal!